src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransTypes.java

```diff
@@ -485,6 +518,11 @@ public void visitMethodDef(JCMethodDecl tree) {
 
     public void visitVarDef(JCVariableDecl tree) {
         tree.vartype = translate(tree.vartype, null);
+        if (tree.init != null) {
+            if (types.isNonNullable(tree.type) && !types.isNonNullable(tree.init.type)) {
```

Comment:
I see why you did this. The synthetic casts are added by `TransTypes` in the various visitor methods. This makes sense because the casts are only required for method calls and field accesses.

My feeling is that most of this remains true also for null checks. E.g. you need to insert a check when you are translating a _variable_ (field, local, method parameter), or a _method call_ - in these cases there might be some nullness markers (on the variable, or on the method return) that should be taken into account, and compared with the "expected" nullness. If a "nullness" mismatch is found, a null-check should be inserted.

So I wonder if the main `translate` method should go from this:

```
translate(expr, expectedErasedType)
```

To this:

```
translate(expr, expectedErasedType, expectedNullness)
```

And the conversions "pushed" down in the visitor, in a similar way as what we do today for synthetic casts.